kernel/kernel_drivers.go

```diff
@@ -165,48 +187,145 @@ func createModulesSubtree(origin, dest string) error {
+	// something goes wrong. Note that this folder needs to be in the same
+	// filesystem as the final one so we can atomically switch the folders.
+	ksnapDir := ksnapName + "_tmp"
+	if opts.KernelInstall {
+		ksnapDir = ksnapName
+	}
 	// Initial clean-up to make the function idempotent
-	if rmErr := RemoveKernelDriversTree(ksnapName, rev); rmErr != nil &&
+	if rmErr := RemoveKernelDriversTree(ksnapDir, rev); rmErr != nil &&
 		!errors.Is(err, fs.ErrNotExist) {
 		logger.Noticef("while removing old kernel tree: %v", rmErr)
 	}
 
-	// Remove tree in case something goes wrong
 	defer func() {
-		if err != nil {
-			if rmErr := RemoveKernelDriversTree(ksnapName, rev); rmErr != nil &&
-				!errors.Is(err, fs.ErrNotExist) {
-				logger.Noticef("while cleaning up kernel tree: %v", rmErr)
-			}
+		// Remove on return if error or if temporary tree
+		if err == nil && opts.KernelInstall {
+			return
+		}
+		if rmErr := RemoveKernelDriversTree(ksnapDir, rev); rmErr != nil &&
+			!errors.Is(err, fs.ErrNotExist) {
+			logger.Noticef("while cleaning up kernel tree: %v", rmErr)
 		}
 	}()
 
-	// Root of our kernel tree
-	treeRoot := driversTreeDir(ksnapName, rev)
+	treeRoot := driversTreeDir(ksnapDir, rev)
 
-	if err := createModulesSubtree(kernelMount, treeRoot); err != nil {
+	// Create drivers tree
+	kversion, err := KernelVersionFromModulesDir(kernelMount)
+	if err == nil {
+		if err := createModulesSubtree(kernelMount, treeRoot,
+			kversion, kmodsInfos); err != nil {
+			return err
+		}
+	} else {
+		// Bit of a corner case, but maybe possible. Log anyway.
+		// TODO detect this issue in snap pack, should be enforced
+		// if the snap declares kernel-modules components.
+		logger.Noticef("no modules found in %q", kernelMount)
+	}
+
+	fwDir := filepath.Join(treeRoot, "lib", "firmware")
+	if opts.KernelInstall {
+		// symlinks in /lib/firmware are not affected by components
+		if err := createFirmwareSymlinks(kernelMount, fwDir); err != nil {
+			return err
+		}
+	}
+	updateFwDir := filepath.Join(fwDir, "updates")
+	// This folder needs to exist always to allow for directory swapping
+	// in the future, even if right now we don't have components.
+	if err := os.MkdirAll(updateFwDir, 0755); err != nil {
 		return err
 	}
+	for _, kmi := range kmodsInfos {
+		if err := createFirmwareSymlinks(kmi.MntDir, updateFwDir); err != nil {
+			return err
+		}
+	}
+
+	// Sync before returning successfully (install kernel case) and also
+	// for swapping case so we have consistent content before swapping
+	// folder.
+	syscall.Sync()
+
+	if !opts.KernelInstall {
+		// There is a (very small) chance of a poweroff/reboot while
+		// having swapped only one of these two folders. If that
+		// happens, snapd will re-run the task on the next boot, but
+		// with mismatching modules/fw for the installed components. As
+		// modules shipped by components should not be that critical,
+		// in principle the system should recover.
```

Comment:
how tights are firmware and modules? would there be a reason that swapping first fimrware would be slightly better?